api/paidAction/index.js

```diff
@@ -107,283 +202,285 @@ export default async function performPaidAction (actionType, args, context) {
+        console.log('enough fee credits available, performing internal action')
+        if ((paymentMethod === 'REWARD_SATS' || paymentMethod === 'FEE_CREDIT')) {
+          await tx.user.update({
+            where: {
+              id: me?.id ?? USER_ID.anon
+            },
+            data: paymentMethod === 'REWARD_SATS'
+              ? { msats: { decrement: cost } }
+              : { mcredits: { decrement: cost } }
+          })
+          console.log('action paid with ' + paymentMethod)
+        } else throw new Error('invalid payment method ' + paymentMethod)
+      } else {
+        console.log('no cost paid action')
+      }
 
-    const result = await action.perform(args, context)
-    await action.onPaid?.(result, context)
+      const result = await performAction(null, paidAction, args, subContext)
+      await paidAction.onPaid?.({ ...result, invoice: null }, subContext)
 
-    return {
-      result,
-      paymentMethod
+      return {
+        result,
+        paymentMethod,
+        retriableWithWallet: false
+      }
+    } catch (e) {
+      console.error('action failed', e)
+      throw e
     }
-  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
+  }
+  const result = context.tx
+    ? await run(context.tx)
+    : await models.$transaction(run, { isolationLevel: Prisma.TransactionIsolationLevel.Serializable })
 
-  // run non critical side effects in the background
-  // after the transaction has been committed
-  action.nonCriticalSideEffects?.(result.result, context).catch(console.error)
+  // run non critical side effects in the background after the transaction has been committed
+  paidAction.nonCriticalSideEffects?.(result.result, context).catch(console.error)
   return result
 }
 
-async function performOptimisticAction (actionType, args, context) {
-  const { models } = context
-  const action = paidActions[actionType]
-
-  context.optimistic = true
-  const invoiceArgs = context.invoiceArgs ?? await createSNInvoice(actionType, args, context)
-
-  return await models.$transaction(async tx => {
-    context.tx = tx
-
-    const invoice = await createDbInvoice(actionType, args, context, invoiceArgs)
-
+async function performOptimisticAction (invoiceData, actionType, paidAction, args, context) {
+  const { models, actionId, me } = context
+  const performInvoicedAction = async tx => {
+    const subContext = { ...context, tx, optimistic: true }
+    await checkUser(me?.id ?? USER_ID.anon, subContext)
+    const invoiceEntry = await createDbInvoice(invoiceData, {
+      actionType,
+      actionId,
+      optimistic: true,
+      args
+    }, subContext)
     return {
-      invoice,
-      result: await action.perform?.({ invoiceId: invoice.id, ...args }, context),
-      paymentMethod: 'OPTIMISTIC'
+      invoice: invoiceEntry,
+      result: await performAction(invoiceEntry, paidAction, args, subContext),
+      paymentMethod: 'OPTIMISTIC',
+      retriableWithWallet: invoiceData.retriable
```

Comment:
when retriableWithWallet is false, it means we have tried all the receiving wallets, this doesn't enforce anything, it is just a flag that lets the clientside code know it should probably use another methods (eg. prioritizeInternal or qr code). The user can still retry if he wishes so, using the retry button.